pandas/core/index.py

```diff
@@ -1548,8 +1548,16 @@ def sym_diff(self, other, result_name=None):
         other, result_name_update = self._convert_can_do_setop(other)
         if result_name is None:
             result_name = result_name_update
-        the_diff = sorted(set((self.difference(other)).union(other.difference(self))))
-        return Index(the_diff, name=result_name)
+        the_diff_sorted = sorted(set((self.difference(other)).union(other.difference(self))))
+        if isinstance(self, MultiIndex):
+            # multiindexes don't currently work well with _shallow_copy & can't be supplied a name
+            return Index(the_diff_sorted, **self._get_attributes_dict())
+        else:
+            # convert list to Index and pull values out - required for subtypes such as PeriodIndex
+            diff_array = Index(the_diff_sorted).values
```

Comment:
...continuing from the previous discussion.

The issue occurred with the subtypes such as `PeriodIndex` where the items that `set` or `sorted` return aren't the values that can go into `_shallow_copy`.

```python
>>> diff
PeriodIndex(['2000-01-03', '2000-01-14'], dtype='int64', freq='B')
>>> set(diff)
set([Period('2000-01-03', 'B'), Period('2000-01-14', 'B')])
```
...but `_shallow_copy` requires the ordinals.

My synthesis on the overall problem is what I said above - that the differences in how some of the subtypes operate cause some issues with unifying the methods. I don't know at what level we'd need to make changes to make this seamless - e.g. we could allow `Period`s to be passed into `PeriodIndex`'s `_shallow_copy` / or have a `_shallow_copy_or_recast`, but is this just pasting over deeper issues